File: java/2272-substring-with-largest-variance.java
class Solution {
    public int largestVariance(String s) {
        var ans = 0;
        for(char maxChar = 'a'; maxChar<='z'; maxChar++) {
            for(char minChar = 'a'; minChar<='z'; minChar++) {
                if (maxChar == minChar) {
                    continue;
                }

                int l = 0;
                int leftRes = 0;
                int rightRes = 0;
                int leftMin = 0;
                boolean hasMin = false;

                for(int r =0 ; r<s.length(); r++) {
                    if (s.charAt(r) == maxChar) {
                        rightRes++;
                    }
                    
                    if (s.charAt(r) == minChar) {
                        rightRes--;
                        hasMin = true;

                        while(l < r) {
                            if (s.charAt(r) == maxChar) {
                                leftRes++;
                                leftMin = Math.min(leftMin, leftRes);
                            }

                            if (s.charAt(r) == minChar) {
                                leftRes--;
                                leftMin = Math.min(leftMin, leftRes);
                            }
                            
                            l++;
                        }
                    }
                    
                    if(rightRes > 0 && hasMin) {
                        ans = Math.max(ans, rightRes - leftMin);
                    }
                }

            }
        }
        return ans;
    }
}

File: python/2272-substring-with-largest-variance.py
class Solution {
    public int largestVariance(String s) {
        Set<Character> uniques = new HashSet<>();
        int[] freq = new int[26];
        
        for (char c : s.toCharArray()) {
            uniques.add(c);
            freq[c-'a']++;
        }
        int max = 0;
        for (char a : uniques) {
            for (char b : uniques) {
                
                int countA = 0, countB = 0, freqA = freq[a-'a'], freqB = freq[b-'a'];
                if (a == b || freqA == 0 || freqB == 0) continue;
                
                for (char c : s.toCharArray()) {
                   
                    if (a == c) 
                        countA++;
                    else if (b == c) {
                        countB--;
                        freqB--;
                    }
                    
                    if (countA + countB < 0 && freqB > 0) {
                        //current total is negative and it would be optimal to start at a A
                        //but we cannot do if no B left.
                        countA = 0;
                        countB = 0;
                    }
                    
                    if (countA != 0 && countB != 0)
                        max = Math.max(max, countA + countB);
                    
                }
            }
        }
        return max;

    }
}